### PR TITLE
Python3 Tool Update

### PR DESCRIPTION
**Pull request by gbalke**
Update the tools to use python3 as python2 has been deprecated for some time. In addition, add some stability fixes.

### CODE REVIEW

File: tools/read_sensor.py [hunk 1]
```diff
@@ -74,13 +72,15 @@
...
                 val = struct.unpack(decode, responses[i])
                 bid = board_ids[i]
                 print("Board:", bid, message.format(args.sensor , val))
-        except ProtocolError as err:
-            print(err)
-            pass
-        except struct.error as err:
+
+        except (MalformedPacketError, ProtocolError) as err:
             print(err)
```
Comment (wuphilipp) [inline tools/read_sensor.py:84]: nit: its `execpt ... as e` in a lot of other places I see, but `err` here

File: tools/comms.py [hunk 2]
```diff
@@ -283,9 +291,12 @@ def eraseFlashSector(self, server_id, sector_nums):
...
         for i in range(0, len(data), COMM_SINGLE_PROGRAM_LENGTH):
-            success = self._programFlashLimitedLength(server_id, dest_addr + i, data[i:i+COMM_SINGLE_PROGRAM_LENGTH])
-            if not success:
-                return False
+            success = False
+            while not success:
+                try:
+                    success = self._programFlashLimitedLength(server_id, dest_addr + i, data[i:i+COMM_SINGLE_PROGRAM_LENGTH])
+                except (MalformedPacketError, ProtocolError) as e:
+                    print(f'Error: Retrying upload of segment {i} of {len(data)}')
```
Comment (wuphilipp) [inline tools/comms.py:299]: I see we are using f strings, which I like! Should we add a todo/issue to convert all of the other `.format` into f strings for consistency? (happy to help with that)

File: tools/upload_firmware.py [hunk 3]
```diff
@@ -39,10 +39,17 @@
...
+                        print(f'Upload to board {board_id} failed with error:')
+                        print(e)
+                        print('Retrying...')
+                        continue
 
                 if success:
-                    # client.leaveBootloader(board_id)
                     print("Upload to Board", board_id, "Succeeded")
                 else:
                     print("Upload to Board", board_id, "Failed")
```
Comment (wuphilipp) [inline tools/upload_firmware.py:55]: it seems like this will never reach the else statement anymore?

File: tools/view_control_loop.py [hunk 4]
```diff
@@ -43,33 +43,28 @@
...
-                data.append(read)
-            except (ProtocolError, struct.error):
-                #print("Failed to communicate with board: ", board_id)
-                data.append([0.0])
-                data.append([0.0])
-        return time.time(), data
+                data.append(struct.unpack('<f', client.readRegisters([board_id], [0x3020], [1])[0]))
+            except (MalformedPacketError, ProtocolError):
+                print("Failed to communicate with board: ", board_id)
+        return time.time(), None if len(data) != 2 else data
```
Comment (wuphilipp) [inline tools/view_control_loop.py:61]: This function looks to be updating the currents for all boards, so we should expect 2 * (number of boards) in the data list?
```suggestion
        return time.time(), None if len(data) != 2 * len(board_ids) else data
```

**Review by wuphilipp — APPROVED** [review]
Looks good! Added some minor nits.

File: tools/comms.py (hunk 2 above)
Comment (gbalke) [reply tools/comms.py:299]: Yes please! Any help would be much appreciated.

File: tools/upload_firmware.py (hunk 3 above)
Comment (gbalke) [reply tools/upload_firmware.py:55]: Yup. I'm debating whether to add a retry limit though so idk.

**Pull request merged**